### PR TITLE
Feature addons, additional functions for simpler steps, `mco_await`, `mco_start`, and `mco_active`

### PR DESCRIPTION
**Pull request by TheTechsTech**
I found some feature in https://github.com/higan-emu/libco project I liked, but some issues https://github.com/higan-emu/libco/pull/14, https://github.com/higan-emu/libco/issues/5 that they decided not to  implement. Which is implemented in this project, that this PR further extends.

Currently they are no functionally changes, just addons that reduce steps. 
- `mco_await` makes any function coroutine using function definition without needing `mco_coro` struct setup and passing.
- `mco_awaitable()` as an preset function to call that executes `mco_await` callable.

The public API now has:

```c
enum value_args_t {
    MCO_FUNC,
    MCO_OBJ,
    MCO_INT,
    MCO_FLOAT,
    MCO_DOUBLE,
    MCO_BOOL,
    MCO_CHR,
    MCO_ARRAY
};

typedef struct mco_value mco_value_t;
typedef struct mco_closure {
    void (*fn)(void *any, ...);
    void *args;
} mco_closure_t;

typedef union {
    mco_closure_t function;
    void *object;
    int integer;
    float point;
    double precision;
    unsigned char boolean;
    char *chars;
    char **array;
}  value_t;

struct mco_value {
  value_t value;
  int type;
  size_t size_args;
  size_t number_args;
};

/* Initialize and starts the coroutine passing persist data and args. */
MCO_API mco_coro *mco_start(void (*func)(mco_coro *co), void *data, void *args);

MCO_API mco_coro *mco_await(void (*func)(void *any, ...), void *args, ...);
MCO_API value_t mco_value(void *data);
MCO_API mco_result mco_suspend(void);
MCO_API mco_result mco_wait(void);
MCO_API mco_coro *mco_active(void);
```

Still needs further development and testing.

```c
MCO_API void mco_send(void *data);
MCO_API value_t mco_receive(void *data);
```

API's are tested in `value_args.c` and `simple_args.c` which was from https://github.com/higan-emu/libco now modified.

Plan on adding GitHub Actions CI tests.

### CODE REVIEW

**Comment by edubart** [comment]
What you are doing here looks like a higher level C wrapper around minicoro functions. In my view such a feature can live fine outside minicoro in your project, as an extended easier to use API for your use case, many people using minicoro this way already.

I don't see this going forward to be merged because of the following points:
- Minicoro was designed to also be usable in other programming language other than C, and the new APIs introduced use of  C varargs (`...`) which are not friendly when binding in other languages.
- There is redundancy in the new introduced functions  `mco_suspend` and `mco_wait`, it's perfectly fine to call `mco_yield(mco_active())`, while it uses more characters, less functions we introduce in the API the simpler it becomes to fit in our head and to bind in other languages.
- There is also divergence from the Lua coroutines API design, which this library tries to mimic. Lua has no concept of coroutine "await".
- The idea of passing values through `value_t` is more limited and less generic than the usual push/pop way, which accepts any kind of C type, even complex structs.
- This idea is making extra temporary allocations, in the performance standpoint this is not good. Minicoro was designed to be efficient and not do allocations behind your back. Making an allocation every "await" call is not ideal, and can have a huge impact on performance.
- These APIs would be even not usable in systems which don't have the concept of general purpose memory allocator, such as freestanding systems, which minicoro should work on. Because it's using `malloc`/`free` calls.
- There are some other issues with the implementation, but I am not going to do a detailed review since I don't see this going forward.
- Finally, minicoro lower level APIs can already cover these use cases, just it's more boilerplate. But again, I encourage people making their own wrappers for their specific use case.

In conclusion the many constraints and goals of this library, doesn't allow higher level C APIs than it already is.
But interesting work! I can see this being used fine in your personal projects.

I did like the CI testing, I could accept a PR with just that.

**Comment by TheTechsTech** [comment]
Your right, it's high level, not much needed to do so. That's the reason I'm using it for the base for **just another programming language**, no need to write everything from scratch. 

All of it is addon at the moment, there is nothing actually fixed or really changing core yet. There will be a lot of breaking changes going forward, this is a playground for a bunch of things. I also was just really seeing where the maintainer mindset is at, quiet a few library projects like this start on high note then go dead. 

I heard about this library from https://news.ycombinator.com/item?id=35133095, I actually tried a few others. 
The concept of `coroutine` is the same, I have an PHP version [symplely/coroutine](https://github.com/symplely/coroutine).

I will close this, and add PR for GitHub Actions CI. Which I now see there is already an actual performance issue, don't see how you can make any comparison claims, without using a standard test, calling there necessary function.

What did you use to compile `benchmark.c` for Windows you have in **README.md**? 
I get error with `_rdtsc`, same for ARM.

**Comment by edubart** [comment]
> Which I now see there is already an actual performance issue, don't see how you can make any comparison claims, without using a standard test, calling there necessary function.

Well I know from experience that calling `malloc()` is bad for performance, it's a block box, it can take an unpredictable amount of time to complete. This is not something I need to test with benchmarks, benchmarking anything with many `malloc()` calls is a bad idea, because you are introducing noise from the system's general purpose allocator in the benchmark. A `malloc()` call can cost different values on Windows, Linux, or depending on libc, and many other environment factors, it's totally a blackbox. Any code that performs `malloc()` is way less deterministic in amount of time to complete, unless you use a custom malloc implementation.

> What did you use to compile benchmark.c for Windows you have in README.md?

The benchmark file was not really designed to be cross platform, so you might have issues. Doing benchmarks in Linux x86_64 using GCC to compile was enough for me to generate data to compare performance and show results in  readme. Of course performance will vary between platform, architecture and OS. You should read those numbers as a hint but not as truth.

> I get error with _rdtsc, same for ARM.

`rdtsc` is an assembly instruction only available in x86 that retrieves the current clock cycle, not sure what replaces it on ARM.  While I've put effort in making minicoro portable to many platforms, I did not do the same for that benchmark file. Do some research and maybe you can come up with a PR fixing `rdtsc` for your platform.

**Comment by TheTechsTech** [comment]
I was referencing to my `to_coro_timing.c` performance test. Now `coro_timing.c` in other PR.
